### ISSUE TITLE
Mutual SSL with OpenSSL

### ISSUE THREAD

**Issue by doom369** [netty/netty#3707]
Hi guys,

I tried to use OpenSSL provider for mutual SSL auth and getting error : 

```
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLException: error:140890B2:SSL routines:SSL3_GET_CLIENT_CERTIFICATE:no certificate returned
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:229)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:339)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:324)
	at io.netty.handler.timeout.ReadTimeoutHandler.channelRead(ReadTimeoutHandler.java:150)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:339)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:324)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:847)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:618)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:329)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:250)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.net.ssl.SSLException: error:140890B2:SSL routines:SSL3_GET_CLIENT_CERTIFICATE:no certificate returned
	at io.netty.handler.ssl.OpenSslEngine.unwrap(OpenSslEngine.java:590)
	at io.netty.handler.ssl.OpenSslEngine.unwrap(OpenSslEngine.java:660)
	at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1114)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:981)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:934)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:315)
	... 13 more
```

With JDK provider everything seems fine, with OpenSSL provider without mutual SSL also everything is working. Does OpenSSL provider mutual auth requires some additional configuration? 
My code is almost same as in JdkSslEngineTest.class specially in mySetupMutualAuth method.

**Comment by normanmaurer** [comment]
@doom369 can you show me the code ?

**Comment by doom369** [comment]
Sure.

Server : 

```
    private static SslContext initSslContext(String serverCertPath, String serverKeyPath, String serverPass,
                                             String clientCertPath,
                                             SslProvider sslProvider) {
        try {
            if (sslProvider == SslProvider.OPENSSL) {
                log.warn("Using native openSSL provider.");
            }

            File serverCert =  new File(serverCertPath);
            File serverKey = new File(serverKeyPath);
            File clientCert =  new File(clientCertPath);

            if (!serverCert.exists() || !serverKey.exists() || !clientCert.exists()) {
                log.warn("ATTENTION. Certificate path not valid. Using embedded certs. This is not secure. Please replace it with your own certs.");
                SelfSignedCertificate ssc = new SelfSignedCertificate();
                return SslContext.newServerContext(sslProvider, ssc.certificate(), ssc.privateKey());
            }

            return SslContext.newServerContext(sslProvider, clientCert, null, serverCert, serverKey, serverPass,
                    null, null, IdentityCipherSuiteFilter.INSTANCE, null, 0 ,0);
        } catch (CertificateException | SSLException | IllegalArgumentException e) {
            log.error("Error initializing ssl context. Reason : {}", e.getMessage());
            throw new RuntimeException(e.getMessage());
        }
    }

 protected void initChannel(SocketChannel ch) throws Exception {
        ChannelPipeline pipeline = ch.pipeline();

        pipeline.addLast(new ReadTimeoutHandler(appTimeoutSecs));

        if (sslCtx != null) {
            SSLEngine engine = sslCtx.newEngine(ch.alloc());
            engine.setUseClientMode(false);
            engine.setNeedClientAuth(true);
            pipeline.addLast(new SslHandler(engine));
        }

        //non-sharable handlers
        pipeline.addLast(new ClientChannelStateHandler(sessionsHolder));
        pipeline.addLast(new MessageDecoder(stats));
        pipeline.addLast(new MessageEncoder());
}
```


Client:

```
    public AppClient(String host, int port, Random msgIdGenerator, boolean disableAppSsl) {
        super(host, port, msgIdGenerator);
        log.info("Creating app client. Host {}, sslPort : {}", host, port);

        if (!disableAppSsl) {
            //todo think how to simplify with real certs?
            //sslCtx = SslContext.newClientContext(getFileFromResources("/test.crt"));
            try {
                //this.sslCtx = SslContext.newClientContext(InsecureTrustManagerFactory.INSTANCE);
                this.sslCtx = SslContext.newClientContext(SslProvider.JDK, new File("/home/doom369/server/certs/server/server.crt"), null,
                        new File("/home/doom369/server/certs/client/client.crt"), new File("/home/doom369/server/certs/client/client.pem"), "my_pass",
                        null, null, IdentityCipherSuiteFilter.INSTANCE, null, 0, 0);
            } catch (SSLException e) {
                log.error("Error initializing SSL context. Reason : {}", e.getMessage());
                log.debug(e);
                throw new RuntimeException();
            }
        }
    }
```

**Comment by normanmaurer** [comment]
@doom369 thanks will check.

**Comment by normanmaurer** [comment]
@doom369 also which version? :)

**Comment by doom369** [comment]
@normanmaurer 4.0.27.Final

**Comment by normanmaurer** [comment]
@doom369 working on it...

**Comment by doom369** [comment]
Cool, thanks. I haven't yet a time to investigate by myself.

**Comment by normanmaurer** [comment]
@doom369 fixed via #3732

**Comment by doom369** [comment]
Just checked it with latest SNAPSHOT. It works! Thank you for quick fix!
